### ISSUE TITLE
Strange behavior of the DMC trial energy

### ISSUE THREAD

**Issue by jtkrogel** [QMCPACK/qmcpack#789]
During a production Mira run with the most recent SoA version of the code I noticed large spikes/re-equilibration when identical DMC sections were run in series (no change in timestep).  This behavior appears to relate to how the trial energy is reset and it undesirable as it leads to unnecessary equilibration and waste of data.  I don't recall observing similar behavior in earlier versions of the AoS code (e.g. v3.1.1).

The last four sections of DMC data shown in the figures below use the same timestep (0.005/Ha) and each includes 10 warmup steps.  

Behavior of the local energy.  Spikes are visible at the onset of each DMC section.  This is not visible for individual twists, but it is for the twist average.
![local_energy](https://user-images.githubusercontent.com/24547905/39269716-96252212-48a1-11e8-9b56-d25e5497bfc5.png)

Behavior of the trial energy.  The trial energy is apparently reset in a way that lacks continuity between the identical DMC sections.
![trial_energy](https://user-images.githubusercontent.com/24547905/39269721-9c93237e-48a1-11e8-8b8e-995c24b12741.png)

**Comment by jtkrogel** [comment]
Additionally, the trial energy above appears to be undershooting the local energy in the first visible DMC section (timestep=0.02/Ha) resulting in beyond target growth of the walker population:
![num_walkers](https://user-images.githubusercontent.com/24547905/39270231-6dcb21ac-48a3-11e8-820f-594128821dcb.png)

**Comment by prckent** [comment]
Agreed that this is a bug. Is this the development version or the latest release?

**Comment by ye-luo** [comment]
Try a run with samples/walkers per thread <= 1

**Comment by jtkrogel** [comment]
What will this show?  The runs are not totally cheap, ~1.5 hrs on 8192 Mira nodes.

**Comment by ye-luo** [comment]
It probably will show no bumps and confirms that you are affected by the >1 bug.

**Comment by jtkrogel** [comment]
Very well.  If there are no objections I will submit a new job tomorrow.

**Comment by jtkrogel** [comment]
I do not set reconfiguration either way.

**Comment by ye-luo** [comment]
Here in the first section, it takes about 100-150 steps to converge and the later sections needs 50 steps. You need to set your warmupSteps close to that. The trial energy is determined as the current localenergy population average during warmup and the history average of the population averaged value after the warmup. So the trial energy of your first section is very much biased due to a small warmup.

From the second to the last block, there is always a bump. I guess that is affected by the nwalker>nthread issue.

**Comment by jtkrogel** [comment]
The trial energy should be determined in such a way that an identical run can follow an already equilibrated run with no warmup steps (anything more is wasteful and prone to be gotten wrong).  More broadly, the behavior of the of the followon identical run should not depend on warmupsteps.

Somewhere (overleaf?) lets write out the formulas used to determine the trial energy and then consider issues of continuity.  The procedure that gets identical followon runs right should also give small reequilibration times for small changes to the timestep in subsequent DMC sections.

I've run DMC in the way set out in this particular run for years without issue.

**Comment by prckent** [comment]
Other codes manage to get this right.

**Comment by ye-luo** [comment]
If the first section is equilibrated, then the following ones should take zero warmup. I'm not convinced that the first one is equilibrated from the the plot.

**Comment by ye-luo** [comment]
It seems that the first section has a different time step than the later blocks. So the second section needs some warmup steps to further equilibrate the system based on the smaller time step. From the third, it should allow zero warmup. It seems to me the current bad bump are due the nwalker>nthread problem.

**Comment by jtkrogel** [comment]
The first two sections are run with a timestep of 0.02/Ha (25 and 400 total steps respectively), the last four use 0.005/Ha (all with 500 total steps).  

The first very short one is not equilibrated upon completion but should be continuous with the second (it is not), and certainly 400 steps at 0.02 timestep is enough to equilibrate (I've run VO2 many times now with a supercell of this size).  

Regarding warmupsteps: a run with 100 warmup steps and 300 non-warmup steps should behave identically to one with 0 warmupsteps and 400 non-warmup steps (perhaps apart from what data is written to scalar.dat), otherwise we always introduce discontinuities at the end of the warmup period as observed here since warmup steps and non-warmup steps are treated differently from the point of view of the trial energy (a discontinuity).

**Comment by jtkrogel** [comment]
Job resubmitted on Mira now with samplesperthread=0.7 and otherwise identical.

**Comment by jtkrogel** [comment]
Following PR #797, discontinuities are no longer visible in the local energy.  Significant discontinuities remain in the trial energy, though it now recovers faster at the start of a new DMC section leading to shorter, but still present, re-equilibration times.  Trial energy discontinuities also remain visible as discontinuities in the walker population.  

See below for plots of these quantities performed as a rerun of the above but with a version of the code following PR #797.

Local energy
![local_energy2](https://user-images.githubusercontent.com/24547905/39535104-14e6eb6c-4e01-11e8-99a1-9e213eb30bf5.png)

Trial energy
![trial_energy2](https://user-images.githubusercontent.com/24547905/39535111-19a08d7a-4e01-11e8-9099-02719dfe0e89.png)

Walker population
![num_walkers2](https://user-images.githubusercontent.com/24547905/39535118-1cf21f70-4e01-11e8-9b5d-54fbce8496e0.png)

**Comment by ye-luo** [comment]
The behaviour of TrialEnergy is expected. 5 warmupsteps for the later sections are added in the input. This triggers the warmup infrastructure assuming the previous section can be anything (VMC, DMC with a different time step...).

**Comment by jtkrogel** [comment]
So is the recommendation to set warmupsteps explicitly to zero in these cases?  Would this eliminate the discontinuities in the trial energy?

It would still be nice if some info from the prior section was forwarded to the following one so that more appropriate action could be taken by the warmup machinery.

**Comment by ye-luo** [comment]
I agree. It needs to be smarter. If the parameters are fully identical between sections, warmup should go zero automatically.
The current behaviour of warmupSteps = 0 is unknown to me.

**Comment by jtkrogel** [comment]
Issues relating to warmup and weight discontinuities remain.
